From aa901d51240db82b5bf8542045d7120c4335f0ff Mon Sep 17 00:00:00 2001
From: Christine Gharzuzi <cgharzuzi@habana.ai>
Date: Thu, 25 Jun 2020 13:43:31 +0300
Subject: [PATCH] [SW-13500] gaudi: fix instability in pci lanes

The amount of time needed to wait between phy init
and executing Synopsys phy W/A differs between HL200 and HL205,
selecting different delays stabilized the first while shook off
the last, this commit removes the delay needed and
replaces it in a monitor loop which waits for indication
from phy (ext_ld_done) which was measured to be around
15us, this changes was tested and proven to be stable
of both HL200 and HL205

Change-Id: I1ddacea45d01eac00b773e08c50f7877f6b46a2f
---

diff --git a/gaudi/common.c b/gaudi/common.c
index 7423a76..c966517 100644
--- a/gaudi/common.c
+++ b/gaudi/common.c
@@ -112,8 +112,12 @@
 	regmap_writel(0x01402801, mmPCIE_DBI_GEN3_RELATED_OFF);
 	regmap_writel(0x04027041, mmPCIE_DBI_GEN3_EQ_CONTROL_OFF);
 
-	/* delay before executing Synopsys W/A > 300us */
-	udelay(576);
+	/* Wait for ext_ld_done before executing W/A */
+	while (regmap_readl(mmPCIE_PHY_PHYX_SRAM_EXT_LD_DONE_0) &&
+		regmap_readl(mmPCIE_PHY_PHYX_SRAM_EXT_LD_DONE_1) &&
+		regmap_readl(mmPCIE_PHY_PHYX_SRAM_EXT_LD_DONE_2) &&
+		regmap_readl(mmPCIE_PHY_PHYX_SRAM_EXT_LD_DONE_3))
+		;
 
 	/* Execute the Synopsys W/A */
 	regmap_writel(0xF, mmPCIE_PHY_CR_PARA_CTRL);
